Use SFTP internally for SCP command

Jakuje opened the pull request:
Based on the discussion on mailing list [1], this is initial implementation of SCP mode using SFTP. Comments, suggestions, reviews, feedback welcomed.

[1] https://lists.mindrot.org/pipermail/openssh-unix-dev/2020-June/038594.html

djmdjm commented on sftp-client.c:
```diff
@@ -1956,3 +1956,52 @@ path_append(const char *p1, const char *p2)
+
+
+int
+is_dir(const char *path)
```
since this will be (more) public, it would probably better as local_is_dir() for clarity

djmdjm commented on sftp-client.h:
```diff
@@ -142,4 +142,17 @@ int upload_dir(struct sftp_conn *, const char *, const char *, int, int, int,
+int is_dir(const char *path);
+
+/* Check whether path returned from glob(..., GLOB_MARK, ...) is a directory */
+int pathname_is_dir(const char *pathname);
```
likewise, s/pathname_is_dir/globpath_is_dir/

djmdjm commented on scp.c:
```diff
@@ -976,19 +1044,37 @@ toremote(int argc, char **argv)
 			addargs(&alist, "%s", host);
-			addargs(&alist, "%s", cmd);
+			/* This will work only if the first remote scp supports sftp mode */
+			addargs(&alist, "%s%s", cmd, mode == MODE_SFTP ? " -M sftp" : "");
```
Will this work? I thought it would treat "cmd -M sftp" as a single argument, though scp's horrible argument passing might avoid that.

Unless there's some reason to do it on a single line, an explicit "if (mode == MODE_SFTP) addargs(...)" would be better IMO

Jakuje replied:
Thank you. Indeed you are right. The arguments should come separately. Fixed now.

djmdjm commented on scp.c:
```diff
@@ -1225,6 +1379,95 @@ rsource(char *name, struct stat *statp)
+
+	remote_path = do_realpath(conn, ".");
+	if (remote_path == NULL) {
+		error("Need cwd");
```
maybe /* TODO - gracefully degrade by using relative paths */ or somesuch here

Jakuje replied:
added

djmdjm commented on scp.c:
```diff
@@ -1225,6 +1379,95 @@ rsource(char *name, struct stat *statp)
+	if (err == -1) {
+		fatal("Failed to download file '%s'", src);
+	}
+	//return(err);
```
remove C++ comment

Jakuje commented:
Thank you for the comments. I addressed your suggestions in the current version.

I will be able to follow-up no earlier than in August on the fallback to scp (thanks for suggesting the mux -- it could actually work that way).

mfriedl commented:
I had a quick first look and I think this looks great!
It also looks like we could put this in soon and keep defaulting to the scp protocol for a while, but
get people to use it and figure out problems.
I'll have a detailed look at the code soon!
thanks!

Jakuje commented:
Updated to resolve conflicts.

I tried also to play with the fallback to scp from sftp, but the current code in sftp-client is really not ready for this as it calls fatal() in `send_msg()` from `do_init()`, which is the place we should just figure out that sftp subsystem request failed so it would require some more tweaking. Putting this aside for now.

Jakuje commented:
I rebased on current master and change default to SFTP to give it more testing. Any update from your side?

mfriedl commented:
I'd like to work on this if spare time permits.

suominen commented on scp.c:
```diff
@@ -526,6 +560,18 @@ main(int argc, char **argv)
+	}
+
+	if (mode == MODE_SFTP && iamremote) {
+		fatal("The server can not be ran in SFTP mode");
```
s/ran/run/ (grammar)

God-knows-all commented:
Please fix it for me. I don’t undetected properly how to fix it? So please
fix it for me.

On Fri, Nov 6, 2020, 6:52 PM Kimmo Suominen <notifications@github.com>
wrote:

> *@suominen* commented on this pull request.
> ------------------------------
>
> In scp.c
> <https://github.com/openssh/openssh-portable/pull/194#discussion_r518731805>
> :
>
> > @@ -526,6 +560,18 @@ main(int argc, char **argv)
>  	/* Do this last because we want the user to be able to override it */
>  	addargs(&args, "-oForwardAgent=no");
>
> +	/* The server can not work in SFTP mode so keep default SCP there */
> +	if (mode == MODE_DEFAULT) {
> +		mode = iamremote ? MODE_SCP : MODE_SFTP;
> +	}
> +	if (mode != MODE_SFTP && sftp_direct != NULL) {
> +		fatal("SFTP direct can be used only in SFTP mode");
> +	}
> +
> +	if (mode == MODE_SFTP && iamremote) {
> +		fatal("The server can not be ran in SFTP mode");
>
> s/ran/run/ (grammar)
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openssh/openssh-portable/pull/194#pullrequestreview-525122500>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANF3EG2JHXVQINWPNZAFE4TSOPWRNANCNFSM4O2OTNQA>
> .
>

DemiMarie commented:
Can we please make quoting bug-compatibility optional?  I really really would like to be able to avoid it in scripts.

djmdjm commented:
All but the switch-default change have been committed (in af5d8094d and 197e29f1c).

My plan is to leave the protocol as scp for at least one release and then look at switching.

Thanks Jakub for writing this!
